test/e2e/federated-service.go

```diff
@@ -106,58 +117,60 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
 			}
 			framework.Logf("%d clusters are Ready", len(contexts))
 
-			clusterClientSets = make([]*release_1_3.Clientset, len(clusterList.Items))
-			for i, cluster := range clusterList.Items {
-				framework.Logf("Creating a clientset for the cluster %s", cluster.Name)
+			// clusters = make([]cluster, len(clusterList.Items))
+			clusters = map[string]*cluster{}
+			primaryClusterName = clusterList.Items[0].Name
+			By(fmt.Sprintf("Labeling %q as the first cluster", primaryClusterName))
+			for i, c := range clusterList.Items {
+				framework.Logf("Creating a clientset for the cluster %s", c.Name)
 
 				Expect(framework.TestContext.KubeConfig).ToNot(Equal(""), "KubeConfig must be specified to load clusters' client config")
 				kubecfg, err := clientcmd.LoadFromFile(framework.TestContext.KubeConfig)
 				framework.ExpectNoError(err, "error loading KubeConfig: %v", err)
 
 				cfgOverride := &clientcmd.ConfigOverrides{
 					ClusterInfo: clientcmdapi.Cluster{
-						Server: cluster.Spec.ServerAddressByClientCIDRs[0].ServerAddress,
+						Server: c.Spec.ServerAddressByClientCIDRs[0].ServerAddress,
 					},
 				}
-				ccfg := clientcmd.NewNonInteractiveClientConfig(*kubecfg, cluster.Name, cfgOverride, clientcmd.NewDefaultClientConfigLoadingRules())
+				ccfg := clientcmd.NewNonInteractiveClientConfig(*kubecfg, c.Name, cfgOverride, clientcmd.NewDefaultClientConfigLoadingRules())
 				cfg, err := ccfg.ClientConfig()
-				framework.ExpectNoError(err, "Error creating client config in cluster #%d", i)
+				framework.ExpectNoError(err, "Error creating client config in cluster #%d (%q)", i, c.Name)
 
 				cfg.QPS = KubeAPIQPS
 				cfg.Burst = KubeAPIBurst
 				clset := release_1_3.NewForConfigOrDie(restclient.AddUserAgent(cfg, UserAgentName))
-				clusterClientSets[i] = clset
+				clusters[c.Name] = &cluster{c.Name, clset, false, nil}
 			}
 
-			clusterNamespaceCreated = make([]bool, len(clusterClientSets))
-			for i, cs := range clusterClientSets {
+			for name, c := range clusters {
 				// The e2e Framework created the required namespace in one of the clusters, but we need to create it in all the others, if it doesn't yet exist.
-				if _, err := cs.Core().Namespaces().Get(f.Namespace.Name); errors.IsNotFound(err) {
+				if _, err := c.Clientset.Core().Namespaces().Get(f.Namespace.Name); errors.IsNotFound(err) {
 					ns := &v1.Namespace{
 						ObjectMeta: v1.ObjectMeta{
 							Name: f.Namespace.Name,
 						},
 					}
-					_, err := cs.Core().Namespaces().Create(ns)
+					_, err := c.Clientset.Core().Namespaces().Create(ns)
```

Comment:
nit: I would prefer us to use f.Namespace.Name here, in preference to 'ns'. I'm not even sure where the latter is declared - I suspect that it's redundant, and should be removed from the framework.  At the very least. the variable used here to create the namespace, and the one used below to report success or failure, should be the same variable.